leptonai/api/v1/api_resource.py

```diff
@@ -0,0 +1,149 @@
+from pydantic import BaseModel
+from requests import Response
+from typing import TYPE_CHECKING, Dict, Union, List, Any, TypeVar, Type, NoReturn
+
+if TYPE_CHECKING:
+    # only used for type hinting, but avoids circular imports
+    from .client import APIClient
+
+
+class ClientError(RuntimeError):
+    def __init__(self, response: Response):
+        super().__init__(
+            f"Client error during API call: {response.status_code} {response.text}"
+        )
+        self.response = response
+
+
+class ServerError(RuntimeError):
+    def __init__(self, response: Response):
+        super().__init__(
+            f"Server error during API call: {response.status_code} {response.text}"
+        )
+        self.response = response
+
+
+class APIResourse(object):
+    """
+    APIResource is a base class for all api implementations. It is registered
+    with the Workspace object and provides a set of utility functions to
+    interact with the API. For example, for all deployment related operations,
+    the DeploymentAPI class is used which is a subclass of APIResource.
+
+    Implementation note: if you are implementing a new set of API, you should subclass
+    APIResource and implement the required methods. And then, in leptonai/api/v1/workspace.py,
+    you should add a new line in the __init__ function to register the new APIResource.
+    For example, if you are implementing a new API for "Magic", you should define a
+    class MagicAPI(APIResource) and then in the __init__ function of Workspace, you should
+    add the following line:
+        self.magic = MagicAPI(self)
+    See for example leptonai/api/v1/deployment.py for an example.
+    """
+
+    _client: "APIClient"
```

Comment:
what is this class variable used for?

Reply:
This is a failsafe approach in case this APIResource needs to call other functionality in the _client class. An example is - if we implement a pod class, then it is likely that we can directly piggyback on _client.deployment. It is not used for now though.